libraries/chain/include/eosio/chain/unapplied_transaction_queue.hpp

```diff
@@ -65,17 +66,25 @@ class unapplied_transaction_queue {
    > unapplied_trx_queue_type;
 
    unapplied_trx_queue_type queue;
+   db_read_mode read_mode = db_read_mode::SPECULATIVE;
```

Comment:
Avoid bringing in controller data structures (like `db_read_mode`) into `unapplied_transaction_queue`. Instead replace both `read_mode` and `only_track_persisted` with something that tracks the add mode for `unapplied_transaction_queue`. Then use a member function to set the add mode at the beginning based on information like the controller read mode and whether any producers are configured on the node.

Here is one option:

Redefine the `trx_enum_type` enum class as:
```cpp
enum class trx_enum_type : uint64_t {
   unknown   = 1,
   persisted = 2,
   forked    = 4,
   aborted   = 8 
};
```
Then add a field into `unapplied_transaction_queue`:
```cpp
    std::underlying_type_t<trx_enum_type> add_mode{};
```

Replace `set_only_track_persisted` and `set_chain_read_mode` with:
```cpp
template<typename... Es>
auto set_add_mode( Es... allowed_types ) -> std::underlying_type_t<trx_enum_type> {
    static_assert( (std::is_same_v<Es, trx_enum_type> && ...) );
    FC_ASSERT( empty(), "set_add_mode, queue required to be empty" );
    return create_mask( allowed_types... );
}

auto set_add_mode() {
    FC_ASSERT( empty(), "set_add_mode, queue required to be empty" );
    return create_mask<trx_enum_type>();
}
```

where the `create_mask` is a utility function added to types.hpp:
```cpp
template<typename E, typename... Es>
inline constexpr auto create_mask( E first, Es... rest )
-> std::enable_if_t<
      std::is_enum_v<E>
      && std::is_integral_v<std::underlying_type_t<E>> 
      && std::is_unsigned_v<std::underlying_type_t<E>>
      && (std::is_same_v<Es, E> && ...),
      std::underlying_type_t<E> 
   >
{
    using F = std::underlying_type_t<E>;
    return (static_cast<F>(first) | ... | static_cast<F>(rest));
}

template<typename E>
inline constexpr auto create_mask() 
-> std::enable_if_t<
      std::is_enum_v<E>
      && std::is_integral_v<std::underlying_type_t<E>> 
      && std::is_unsigned_v<std::underlying_type_t<E>>,
      std::underlying_type_t<E> 
   >
{
    return std::underlying_type_t<E>{};
}
```

Then you check for the appropriate flag set in the `add_mode` member variable in each of the add member functions. For example:
```cpp
void add_aborted( std::vector<transaction_metadata_ptr> aborted_trxs ) {
   if( aborted_trxs.empty() || !has_field( add_mode, trx_enum_type::aborted ) ) return;
   ...
}
```

Reply:
After talking with @heifner: 

The intention is for `unapplied_transaction_queue` to be aware of the logic of when/how transactions get abandoned due to forks, aborts, etc., and to impose the policy for ordering these transactions. So the decision of which add functions do nothing under certain environments is to be made by `unapplied_transaction_queue`. It just needs to know what the relevant environment is to make its decisions.

I still think it isn't good to bring in ` db_read_mode` into this class however. So there can be a new `enum class` that describes the relevant environment and is fed into a single function that sets the mode (i.e. sets the value of member variable appropriately so that future add calls work as intended). As of now, we only need three cases to distinguish for the environment: `speculative_producer`, `speculative_non_producer`, `non_speculative`.